### ISSUE TITLE
Feature Request: ACL variables

### ISSUE THREAD

**Issue by trondhindenes** [hashicorp/consul#3957]
It would be awesome if Consul ACLs supported variables, especially `$self` or something similar pointing to the local agent.

The more I'm looking at ACLs and how to lock them down properly, the more I realize that one (today) either has to:
- dynamically provision a token per agent, or
- allow "all" agents to perform operations on "all other" agents.

I think variable-support would combat at least some of these shortcomings. It would allow ACL expressions such as:
```
node "$self" {
  policy = "write"
}

agent "$self" {
  policy = "write"
}
```
, which would allow expressing an agents' permissions against "itself" without having to implement unique ACL tokens per agent.

Just a suggestion.

**Comment by banks** [comment]
Hey @trondhindenes,

That's an interesting idea, thanks.

I'm not really sure if it works though: if all agent's have the same secret then what is stopping an "bad" agent from trying to register as a different node name? If `$self` is just taken from the node name the agent declares then there is no actual security here - you might as well have a single ACL for all nodes.

The ACL system relies on a different secret per actor as it's the only verifiable bit of identity we have, there aren't really any shortcuts there. We have in mind ways to improve our integration with Vault though which will allow simpler setups that hook into external identity mechanisms with a strong trust root - e.g. AWS IAM instance roles etc. and then automate provision based on that.

**Comment by trondhindenes** [comment]
that's a valid point (and one that should maybe be surfaced more clearly in the docs)
When we started using Consul we decided to keep it simple, and have a unique ACL per service. As such we don't care that much if a client registers with a different name, as long as it doesn't have access to register the "wrong" service (based on service name prefixes in ACL). We don't use the same token for all servers, but each identical group of servers gets the same token in our current setup.

Services are far more touchy than agent names in our environment, which is why a "shortcut" to allow simple-to-use "self" permissions would be awesome.
